go-controller/pkg/ovn/namespace.go

```diff
@@ -0,0 +1,122 @@
+package ovn
+
+import (
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	kapi "k8s.io/client-go/pkg/api/v1"
+	"sync"
+	"time"
+)
+
+func (oc *Controller) waitForNamespaceEvent(namespace string) {
+	for {
+		if oc.namespacePolicies[namespace] != nil {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+func (oc *Controller) addPodToNamespaceAddressSet(ns, address string) {
+	if oc.namespacePolicies[ns] == nil {
+		return
+	}
+
+	oc.namespaceMutex[ns].Lock()
+	defer oc.namespaceMutex[ns].Unlock()
+
+	oc.namespaceAddressSet[ns] = append(oc.namespaceAddressSet[ns],
+		address)
+	oc.setAddressSet(ns, oc.namespaceAddressSet[ns])
+}
+
+func (oc *Controller) deletePodFromNamespaceAddressSet(ns, address string) {
+	if address == "" || oc.namespacePolicies[ns] == nil {
+		return
+	}
+
+	oc.namespaceMutex[ns].Lock()
+	defer oc.namespaceMutex[ns].Unlock()
+
+	for i, item := range oc.namespaceAddressSet[ns] {
+		if item == address {
+			oc.namespaceAddressSet[ns] = append(
+				oc.namespaceAddressSet[ns][:i],
+				oc.namespaceAddressSet[ns][i+1:]...)
+		}
+	}
+
+	oc.setAddressSet(ns, oc.namespaceAddressSet[ns])
+}
+
+func (oc *Controller) addNamespace(ns *kapi.Namespace) {
```

Comment:
I say that in case we do not see the Namespace yet, we still create all of it in a separate table assuming that it will be linked up when the Namespace event finally happens asynchronously (under a lock though).
All such asynchronous threads can just declare that something changed and issue a 'commit' call to a worker queue. And one global worker processes the queue. It can skip a commit call because of insufficient data and will process when the next event arrives in its queue.
Just synchronizing based on events as against the current way of having to wait in a polling loop.

It may be too much to handle it in the first revision itself. So lets discuss this further and attack it in subsequent PRs.